p2p/protocol/identify/obsaddr.go

```diff
@@ -185,20 +230,92 @@ func (oas *ObservedAddrSet) gc() {
 	}
 }
 
-func (oas *ObservedAddrSet) doAdd(observed, local, observer ma.Multiaddr,
-	direction network.Direction) {
+func (oas *ObservedAddrManager) addConn(conn network.Conn, observed ma.Multiaddr) {
+	oas.activeConnsMu.Lock()
+	defer oas.activeConnsMu.Unlock()
+
+	// We need to make sure we haven't received a disconnect event for this
+	// connection yet. The only way to do that right now is to make sure the
+	// swarm still has the connection.
+	//
+	// Doing this under a lock that we _also_ take in a disconnect event
+	// handler ensures everything happens in the right order.
+	for _, c := range oas.host.Network().ConnsToPeer(conn.RemotePeer()) {
+		if c == conn {
+			oas.activeConns[conn] = observed
+			return
+		}
+	}
+}
+
+func (oas *ObservedAddrManager) removeConn(conn network.Conn) {
+	// DO NOT remove this lock.
+	// This ensures we don't call addConn at the same time:
+	// 1. see that we have a connection and pause inside addConn right before recording it.
+	// 2. process a disconnect event.
+	// 3. record the connection (leaking it).
+
+	oas.activeConnsMu.Lock()
+	delete(oas.activeConns, conn)
```

Comment:
log / warn if conn isn't in oas.activeConns ?

Reply:
We call removeConn for every connection, but not all of them will be in this set.

Reply:
That is, not all connections will be associated with a valid observation.